File: marchingBand.py
'''
Solution to MPMP12: MARCHING BAND PROBLEM
https://www.youtube.com/watch?v=2DNokQGxqjE


Output:
    Use 7560 marchers to make 64 options

    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 14, 15, 18, 20, 21, 24, 27, 28, 30, 35, 36, 40, 42, 45, 54, 56, 60, 63, 70, 72, 84, 90, 105, 108, 120, 126, 135, 140, 168, 180, 189, 210, 216, 252, 270, 280, 315, 360, 378, 420, 504, 540, 630, 756, 840, 945, 1080, 1260, 1512, 1890, 2520, 3780, 7560]


'''

import math

target = 64


def numberOfOptions(numOfMarchers):
    # Every option will have one side less than the root of the numOfMarchers
    # Edge case: both sides are equal to the square root
    rootOf = math.sqrt(numOfMarchers)
    options = []
    for i in range(1, math.floor(rootOf) +1):
        if numOfMarchers % i == 0:
            options.append(i)
            if i != rootOf: # check for the edge case
                options.append(int(numOfMarchers /i))
    return len(options), options

i = 1
while True:
    num, options = numberOfOptions(i)
    if  num == target:
        print ("Use %s marchers to make %s options\n"%(i, target))
        options.sort()
        print(options)
        break
    i += 1

File: marchingBand.rs
/*
Solution to MPMP12: MARCHING BAND PROBLEM
https://www.youtube.com/watch?v=2DNokQGxqjE

Output:
    Use 7560 marchers to make 64 options

*/


// Every option will have one side less than the root of the numOfMarchers
// Edge case: both sides are equal to the square root
fn number_of_options(num_of_marchers: i32) -> i32{
	let root_of = (num_of_marchers as f32).sqrt().floor() as i32;

	let mut count = 0;
	for i in 1..(root_of+1) {
		if (num_of_marchers as f32) % (i as f32) == 0.0 {
			count += 1 ;
			if i != root_of { //check for the edge case
				count += 1;
			}
		}
    }
	return count;
}


fn main() {
	let target = 64;
	
	let mut i = 1;
	loop {
		if number_of_options(i) == target {
			println!("Use {} marchers to make {} options\n", i, target);
			break;
		}
		i += 1;
	}
}
